### ISSUE TITLE
Return response instead of injecting it to the responder

### ISSUE THREAD

**Issue by PatrickLouys** [amphp/aerys#109]
The current example doesn't make much sense in my opinion (I did something similar for my own HTTP component which I came to dislike). 

The end method clearly has side effects. What happens if I call it twice? If something happens below it? It is not clear from looking at the API.

```
(new Aerys\Host)->use(function(Aerys\Request $req, Aerys\Response $res) {
	$res->end('foo');
});
```

Instead I suggest the following: 
```
(new Aerys\Host)->use(function(Aerys\Request $req) {
	return new Aerys\Response('foo');
});
```

A request handler takes the request as an argument and returns a response. I think this would be much more sensible.

I have no idea if it is still feasible to change this and maybe you guys had this argument already. Just something that bugged me when looking at the docs.

**Comment by bwoebi** [comment]
It is disallowed to call end() twice and will lead to an exception.

And yes, this is mainly to allow streamed responses. If the complete response would be always available at once, this were a viable approach.

But with streamed responses, you cannot return (or it will be really ugly with an extra callback for emitting further data… ugh.) before you have sent everything and you don't want to buffer.

Also, `$res->end()` trivially allows continuing to process data after submitting the response without an extra coroutine.

Ultimately, except for the simplest case, it's not a very nice solution.

Also, it is recommended to avoid passing the $res Response object around, to avoid the issue you mentioned.

**Comment by kelunik** [comment]
As @bwoebi already mentioned, this is by design to allow streaming responses. In addition, once a coroutine / handler ends, the response is automatically ended. If we had another streaming mechanism via callbacks, we can't guarantee that a handler is still alive and didn't "crash" due to some exception.

**Comment by PatrickLouys** [comment]
I think having a callback would be the nicer solution (Symfony also opted for this with their StreamedResponse).

@bwoebi my SO Chat crashed, so here is the response to your last reply:
which the return would take care of, you build your object and then outside of the handler the response it sent, with everything in the correct order.

I shouldn't have to worry about the order that methods are called as a developer. The API that I am using directly should not have any side effects.

Doing a return with the callback for streaming would take care of all that and provide a clean API without sideeffects. You can make an interface so that people can implement their own Response Implementations (JsonResponse, ...).

I don't understand enough about the topic to know how much of an issue the things are that @kelunik described. I'm sure there could be a workaround for that?

As a user I am mostly worried about having a nice, clean API to work with. In it's current state, I can't consider aerys because of that, which sucks because I really like the idea.

**Comment by kelunik** [comment]
@PatrickLouys You will always have side effects if you allow streaming responses including flushing. I don't see how your proposal removes them. Could you give a more complete usage example?

**Comment by PatrickLouys** [comment]
@kelunik It won't remove them for streaming responses if the Symfony model is followed. But at least now they only exist for streamed responses (in the callback). Otherwise you have to deal with them for all responses.

I haven't used streamed responses more than a couple of times in my career so far, so I don't know too much about the details. I am worried about the API of the things that I use every day.

With end(), how am I supposed to do post-processing with the response before it is sent, caching for example?

**Comment by PatrickLouys** [comment]
@kelunik  How about a Flushable Response that's another Response implementation?

**Comment by bwoebi** [comment]
For post-processing, we have middlewares. (see http://amphp.org/docs/aerys/middlewares/do.html)

For caching you could have a class like:

```php
class Cache implements Aerys\Middleware {
    function do(Aerys\InternalRequest $ireq) {
        $headers = $data = yield;
        $body = "";
        do {
            $data = yield $data;
            $body .= $data;
        } while ($data !== null);
        $this->writeToCache($ireq->uri, $headers, $body);
    }

    function __invoke(Aerys\Request $req, Aerys\Response $res) {
        if (list($headers, $body) = $this->fetchFromCache($req->getUri())) {
            foreach ($headers as $name => $vals) {
                foreach ($vals as $val) {
                    $res->addHeader($name, $val);
                }
            }
            $res->end($body);
        }
        /* otherwise we fallback to the next request handler if not yet in cache */
    }
}
```

**Comment by bwoebi** [comment]
@PatrickLouys How about KISS and just having one single Response implementation? ;-)

That way we'll end up with multiple implementations and it will only make the learning curve steeper…

**Comment by PatrickLouys** [comment]
@bwoebi Strongly disagree. I don't want to reinvent the wheel every time. There should be different Responses (JsonResponse, StreamedResponse, maybe even stuff like TwigResponse...).

SOLID is a thing for a reason :)

**Comment by kelunik** [comment]
@PatrickLouys If you argue with SOLID, what's the difference of `JsonResponse` that would implement `Response`?

**Comment by joshdifabio** [comment]
Streaming request/response support is absolutely necessary. However, it's also worth remembering that in 99% of cases it's not actually needed. I think @PatrickLouys is right in saying that it should be easier to specify routes which accept an immutable request object and return a promise which resolves to an immutable response object (e.g. the PSR7 implementations, even if the stream implementation is crappy). It's absolutely possible to do this using Aerys right now, but it's more difficult than it should be given that it's what people want to do in the vast majority of cases.

**Comment by PatrickLouys** [comment]
@kelunik has content type already set and automatically json_encodes what you pass in the constructor for example.

**Comment by kelunik** [comment]
> In it's current state, I can't consider aerys because of that, which sucks because I really like the idea.

To answer that part at least: You can always use your own small response object and have a generic wrapper for your handlers that reads from the custom response object and passes these on to an Aerys response. It shouldn't be a blocker for you.

**Comment by kelunik** [comment]
I'm reopening this, as we have to rethink it together with #156. Currently it's way too easy to forget a `return;` to end a responder early, e.g.

```php
if (!$hmacSigner->verify($payload, $signature)) {
    $response->setStatus(400);
    $response->end("<h1>Invalid Request</h1>");

    return;
}
```

A response could actually look like that:

```php
if (!$hmacSigner->verify($payload, $signature)) {
    return new Response(400, "<h1>Invalid Request</h1>");
}
```

with

```php
interface Response {
    function getStatus(): int;
    function getReason(): string;
    function getHeaders(): array;
    function getBody(): InputStream;
}
```

Most responses don't need /want to set the reason phrase or use streaming, they can just use an implementation that just accepts a status code, response body as string, and an optional array of 

Streaming responses on the other hand can return a response and continue adding e.g. to an `IteratorStream` outside the coroutine. For real streaming responses (not file streaming via `ByteStream\pipe`) the auto-`end()`-ing of responses can be quite annoying I guess.

One thing that's still missing with that approach is auto-aborting on client disconnect. I think we can just expose a `Client` object as part of the request that can be periodically checked, which is what most streaming servers probably want to do anyway instead of periodically sending a dummy message and waiting for Aerys to throw an exception for cleanup.

**Comment by kelunik** [comment]
@bwoebi @trowski What's your opinion on this?

**Comment by kelunik** [comment]
I'm adding some thoughts I had, just so they don't get lost.

```php
interface Response { // <-- can be a final class, as it doesn't contain any logic
    function getStatus(): int;
    function getReason(): string;
    function getHeaders(): array;
    function getBody(): InputStream;
}

interface Request {
    // ...

    /**
     * This can be used to stop streaming.
     * This will _always_ be called if the response stream has been fully written,
     * even if the connection is kept alive.
     */
    function onClientClose(callable $onClientClose);
}

interface Responder {
    // Some other entity is responsible for building a responder chain.
    // $next would be unused in actual responders, which seems sub-optimal.

    function processRequest(Request $request, Responder $next): Promise<Response>;
}

interface Responder {
    // If the responder wants to delegate to other responders,
    // it can require those to be passed via the constructor.

    function processRequest(Request $request): Promise<Promise>;
}

createHtmlResponse(int $status, array $headers, string $body) { ... }
createJsonResponse(...)

$router = new Router;

$router->route("GET", "/foobar", new SomeMiddleware(new FoobarHandler));
```

**Comment by kelunik** [comment]
Returning responses is implemented as part of the upcoming 0.8 release.
